src/win/fs.c

```diff
@@ -259,57 +304,107 @@ INLINE static void uv__fs_req_init(uv_loop_t* loop, uv_fs_t* req,
 }
 
 
-static int fs__wide_to_utf8(WCHAR* w_source_ptr,
-                               DWORD w_source_len,
-                               char** target_ptr,
-                               uint64_t* target_len_ptr) {
-  int r;
-  int target_len;
-  char* target;
-  target_len = WideCharToMultiByte(CP_UTF8,
-                                   0,
-                                   w_source_ptr,
-                                   w_source_len,
-                                   NULL,
-                                   0,
-                                   NULL,
-                                   NULL);
+static int32_t fs__get_surrogate_value(const WCHAR* w_source_ptr,
+    DWORD w_source_len) {
+  WCHAR u;
+  u = w_source_ptr[0];
+  if (u >= 0xd800 && u <= 0xdbff && w_source_len > 1) {
+    WCHAR next;
+    next = w_source_ptr[1];
+    if (next >= 0xdc00 && next <= 0xdfff)
+      return 0x10000 + ((u - 0xd800) << 10) + (next - 0xdc00);
+  }
+  return u;
+}
 
-  if (target_len == 0) {
-    return -1;
+
+static int fs__get_length_wide(const WCHAR* w_source_ptr, DWORD w_source_len) {
+  int target_len;
```

Comment:
Could you change all of the size arguments to type ssize_t in the new code?
```suggestion
static ssize_t fs__get_length_wide(const WCHAR* w_source_ptr, DWORD w_source_len) {
  ssize_t target_len;
```